libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt

```diff
@@ -792,21 +800,33 @@ class MapboxNavigation @VisibleForTesting internal constructor(
             routes,
             initialLegIndex,
             reason,
+            callback,
         )
     }
 
     private fun internalSetNavigationRoutes(
         routes: List<NavigationRoute>,
         legIndex: Int = 0,
         @RoutesExtra.RoutesUpdateReason reason: String,
+        callback: RoutesSetCallback? = null,
     ) {
         rerouteController?.interrupt()
         restartRouteScope()
         threadController.getMainScopeAndRootJob().scope.launch(Dispatchers.Main.immediate) {
             routeUpdateMutex.withLock {
                 navigationSession.setRoutes(routes)
-                setRoutesToTripSession(routes, legIndex, reason)
-                directionsSession.setRoutes(routes, legIndex, reason)
+                when (val processedRoutes = setRoutesToTripSession(routes, legIndex, reason)) {
+                    is NativeSetRouteValue -> {
+                        directionsSession.setRoutes(routes, legIndex, reason)
+                        callback?.onRoutesSetResult(RoutesSetCallbackSuccess(routes))
+                    }
+                    is NativeSetRouteError -> {
+                        logW("Routes $routes will be ignored as they are not valid")
+                        callback?.onRoutesSetError(
+                            RoutesSetCallbackError(routes, processedRoutes.error)
+                        )
+                    }
+                }
```

Comment:
I just now see that this is already implemented and actually suffers from the issue I mentioned in https://github.com/mapbox/mapbox-navigation-android/pull/5938#discussion_r915770292.

The `NavigationSession.setRoutes(routes)` should not be called if `DirectionsSession.setRoutes(routes, legIndex, reason)` wasn't.

Reply:
Tracking in https://github.com/mapbox/mapbox-navigation-android/issues/6023.